dask/dataframe/tests/test_groupby.py

```diff
@@ -2501,3 +2501,48 @@ def test_empty_partitions_with_value_counts():
     ddf = dd.from_pandas(df, npartitions=3)
     actual = ddf.groupby("A")["B"].value_counts()
     assert_eq(expected, actual)
+
+
+def test_groupby_grouper_not_implemented():
+    # DataFrame
+    pdf = pd.DataFrame(
+        {
+            "a": [1, 2, 3, 4, 5],
+            "b": ["1985", "1985", "1990", "1990", "1990"],
+        },
+    )
+    pdf["b"] = pd.to_datetime(pdf["b"], format="%Y")
+
+    ddf = dd.from_pandas(pdf, npartitions=3)
+
+    with pytest.raises(NotImplementedError):
+        ddf.groupby(pd.Grouper(key="b", freq="y")).apply(
+            lambda x: x["a"].mean(), meta=pd.Series(dtype="float", name="b")
+        ).compute()
+
+    pdf2 = pd.DataFrame(
+        {
+            "Publish date": [
+                pd.Timestamp("2000-01-02"),
+                pd.Timestamp("2000-01-02"),
+                pd.Timestamp("2000-01-09"),
+                pd.Timestamp("2000-01-16"),
+            ],
+            "ID": [0, 1, 2, 3],
+            "Price": [10, 20, 30, 40],
+        }
+    )
+    ddf2 = dd.from_pandas(pdf2, npartitions=3)
+
+    with pytest.raises(NotImplementedError):
+        ddf2.groupby(pd.Grouper(key="Publish date", freq="1W")).mean().compute()
```

Comment:
Does this test something different than the DataFrame groupby above? Naively I was expecting two `with pytest.raises(NotImplementedError):` checks -- one for a DataFrame and one for a Series

Reply:
Yes, indeed. The one above is testing a different case. I'll try to explain via an interactive session excerpt (using `dask-2021.3.0`, i.e., none of the code from this PR is in effect):
```python
# [Case 1] Pandas dataframe
In []: pdf.groupby(pd.Grouper(key="b", freq="y")).apply(                                                  
    ...:     lambda x: x["a"].mean(),         
    ...: )                                                  
Out[]:                                                                                                                                                                                                              
b                                                                              
1985-12-31    1.5                              
1986-12-31    NaN                                                    
1987-12-31    NaN                                                                                             
1988-12-31    NaN                                
1989-12-31    NaN  
1990-12-31    4.0                                                                                                                
dtype: float64                                                                

# [Case 1] Dask dataframe                                  
In []: ddf.groupby(pd.Grouper(key="b", freq="y")).apply(           
    ...:     lambda x: x["a"].mean(), meta=pd.Series(dtype="float", name="b")                   
    ...: ).compute()   
---------------------------------------------------------------------------
...
ValueError: Length of passed values is 0, index implies 2.

# [Case 1] Dask dataframe - but without employing `.apply()`
In []: ddf.groupby(pd.Grouper(key="b", freq="y")).mean().compute()                            
Out[]:                                                                     
a                                                                                             
b                                                                                                 
1985-12-31  1.5                                                            
1990-12-31  4.0 

# [Case 2] Pandas dataframe
In []: pdf2.groupby(pd.Grouper(key="Publish date", freq="1W")).mean()
Out[]: 
               ID  Price
Publish date            
2000-01-02    0.5   15.0
2000-01-09    2.0   30.0
2000-01-16    3.0   40.0

# [Case 2] Dask dataframe
In []: ddf2.groupby(pd.Grouper(key="Publish date", freq="1W")).mean().compute()
Out[]: 
               ID  Price
Publish date            
2000-01-02    0.5   15.0
2000-01-09    2.0   30.0
2000-01-16    3.0   40.0
```

Both of those use `pd.Grouper`. Case (1) was the example provided in the [opening comment](https://github.com/dask/dask/issues/6104#issue-600654236) of the issue.
I came up with case (2) to demonstrate that `pd.Grouper` is not necessarily working incorrectly on Dask's side (in case (2), it does work as expected), but there are some edge cases which still need ironing out. One of which seems to be how `dask.dataframe.groupby.DataFrameGroupBy` objects handle calls to `.apply()` (the circumstantial evidence suggests that there might be some `pd.Grouper` involvement). I haven't had a chance yet to have a proper look at what's going on there.

Please, note that not all calls to `.apply()` exhibit erroneous behavior:
```
# Pandas
In []: rng = np.random.RandomState(8)
    ...: df = pd.DataFrame(
    ...:       rng.randn(1000, 3),
    ...:       index=pd.date_range("1/1/2000", periods=1000),
    ...:       columns=["A", "B", "C"],
    ...: )

In []: 

In []: grouped = df.groupby("A")

In []: grouped["C"].apply(lambda x: x.mean())
Out[]: 
A
-3.818256    0.414952
-3.559460   -0.117778
-3.134920   -1.869799
-2.886224   -0.364857
-2.851524   -1.518744
               ...
 2.653087   -0.612375
 2.810955    1.725185
 2.895935   -0.593584
 2.900929    0.465982
 3.060468    0.977948
Name: C, Length: 1000, dtype: float64

# Dask
...
In []: grouped["C"].apply(lambda x: x.mean()).compute()
<ipython-input-65-3c6bef503e97>:1: UserWarning: `meta` is not specified, inferred from partial data. Please provide `meta` if the result is unexpected.
  Before: .apply(func)                                                     
  After:  .apply(func, meta={'x': 'f8', 'y': 'f8'}) for dataframe result
  or:     .apply(func, meta=('x', 'f8'))            for series result
  grouped["C"].apply(lambda x: x.mean()).compute()
Out[]:                                                                                                                
A                                                                                
-3.818256    0.414952
-2.395213    1.688095                                          
-2.349397   -0.528059                                 
-2.067072    0.720787
-1.996921    0.120153
               ...                                                                      
 2.515079    0.031373                           
 2.810955    1.725185                           
 2.895935   -0.593584                                                           
 2.900929    0.465982
 3.060468    0.977948         
Name: C, Length: 1000, dtype: float64
```

That's why my understanding has been that - for the time being, until we get to the bottom of what discrepancies exist `pd.Grouper`-wise - as an interim step, the support of `pd.Grouper` will be removed (as per the most recent two comments in the issue thread, namely [this one](https://github.com/dask/dask/issues/6104#issuecomment-617784303) and [this one](https://github.com/dask/dask/issues/6104#issuecomment-797525562)).

Given this further explanation, what are your thoughts on how it'd be best to proceed? Should work on this be suspended (effectively, removing the interim step) and instead focusing solely on figuring out what the root cause of  the underlying `pd.Grouper` issue(s) in Dask is?

Or should this still proceed (with the caveat that it'll be removing support for some legitimately working cases; probably, undesirable) with this and remove the `.compute()` operations as per your suggestions and add some comments to explain why there are two tests and _what_ they are testing and demonstrating, in particular.

Reply:
The second code block has been updated with some further clarifications.

Reply:
Thanks @hristog -- FWIW I don't see the updates in the PR diff. Perhaps there are commits that still need to be pushed?

Reply:
Apologies, @jrbourbeau - I should've probably been clearer.
I haven't performed any code changes yet. I'm happy to remove the `.compute()` calls from the tests - it does make sense, since - as you've said - the exception gets raised at an earlier stage.

I've updated the [actual comment](https://github.com/dask/dask/pull/7410#discussion_r599147756) (the second code block in it, and the associated explanation immediately above) to demonstrate the degree of what's going wrong. As you can see, one possible explanation is that `.apply()` is not behaving correctly in all cases (even ones which have nothing to do with `pd.Grouper`, in particular).

In order to proceed with further code changes, I'll probably need some input regarding the outstanding questions, as they pertain to how this should be moved further:
> Given this further explanation, what are your thoughts on how it'd be best to proceed? Should work on this be suspended (effectively, removing the interim step) and instead focusing solely on figuring out what the root cause of  the underlying `pd.Grouper` issue(s) in Dask is?

> Or should I still proceed (with the caveat that it'll be removing support for some legitimately working cases; probably, undesirable) with this and remove the `.compute()` operations as per your suggestions and add some comments to explain why there are two tests and _what_ they are testing and demonstrating, in particular?

Reply:
An attempt to summarize the different use/test cases from my [comment above](https://github.com/dask/dask/pull/7410#discussion_r599147756).
| Grouping use case | Matches Pandas behavior? |
|:-------------|:-------------|
|`.groupby("A")["C"].apply(lambda x: x.mean(), meta=pd.Series(dtype="float")).compute() # for npartitions=1`| Yes |
|`.groupby("A")["C"].apply(lambda x: x.mean(), meta=pd.Series(dtype="float")).compute() # for npartitions>1`| No (different output) |
|`.groupby(pd.Grouper(key="b", freq="y")).mean().compute()`       | Almost (the difference is that it doesn't output `NaN`s) |
|`.groupby(pd.Grouper(key="b", freq="y")).apply(lambda x: x["a"].mean())`| No (raises `ValueError`) |
|`.groupby(pd.Grouper(key="Publish date", freq="1W")).mean()`|Yes|

Reply:
@jrbourbeau checking in here, if I'm not reading incorrectly @hristog wants some input on this https://github.com/dask/dask/pull/7410#discussion_r599783028 to know how to move forward. Would you be able to provide some feedback or ping someone that can?

Reply:
Hi @ncclementi - thanks for bringing this up! I confirm that any kind of input from the maintainers team on this PR would be greatly appreciated :)

Reply:
#7857 already raises NotImplementedError, so the remaining work is to try to add support for pd.Grouper. I am going to close this PR, but you should open a new one if you'd like to start implementing that.

Reply:
Thanks for confirming, @jsignell! Indeed, I did see the changes in #7857, when that PR was merged in.